### ISSUE TITLE
Issue with license file checkout when using singularity exec, but not with singularity shell?

### ISSUE THREAD

**Issue by scottnsbtt** [sylabs/singularity#520]
**Version of Singularity**
3.8.5

**Type of issue**
We are running Abaqus 2019 in a singularity container on an HPC using the PBS batch script with the PBS job scheduler. When running the job, there is no issue with Abaqus checking out its license from the remote license server (possibly because we specify the path to the license server inside an "abaqus_v6.env" file located in the working directory), but there are two other software we use in conjunction with Abaqus (one is the Intel fortran compiler (Comp-FL), the other is a third party abaqus plugin) and they are not acting as expected when checking out their licenses. The license check-out behaviour changes depending on whether we use "singularity exec" via a PBS batch script or "singularity shell". The license manager is flexlm in all cases.

The locations of the license files are set in the pbs batch script using the following lines prior to invoking singularity
```
###set environment variable for Intel Fortran Compiler license file
export SINGULARITYENV_INTEL_LICENSE_FILE="[port number redacted]@[server name redacted]"
###set environment variable for the third party plugin license file
export SINGULARITYENV_PLUGIN_LICENSE_FILE="[port number redacted]@[server name redacted]"
```

When running the job via PBS job scheduler and batch script containing "singularity exec", we get errors about not being able to checkout the licenses for Comp-FL (Intel compiler) and the third party plugin. This is despite the fact that we do see the correctly set environment variables when we run "PRINTENV" via "singularity exec", so we do know that these environment variables are being set and passed into the container properly.

When running the job without PBS scheduler by setting the environment variables and simply running "singularity shell" on the login node of the HPC and then running Abaqus directly from the singularity shell we have no issue checking out the license for the third party plugin  (i.e. I get the desired result, which is different than with "Singularity exec"). Despite this, I still cannot get the Fortran license to checkout even from within a singularity shell (this might be a separate issue to do with the Intel Compiler suite).

Running "lmstat -a -c [portnumber]@[server name redacted]" on the host HPC as well as within the singularity container shell on the host HPC both return a good result that clearly shows all the licenses being available (i.e. this is not a network issue affecting the connection to the license server). However, running "lmstat -a -c [portnumber]@[server name redacted]" via "singularity exec" via a PBS batch script returns "INTEL: Cannot connect to license server system. (-15,570:115 "Operation now in progress")"

TLDR: License checkout result is not the same when using "singularity exec" via a pbs batch script run through the PBS job scheduler vs "singularity shell" run on the login node of the HPC (no PBS batch scheduler involvement)

**error messages**
***for Intel Fortran Compiler***
The following is error output when running an abaqus job requiring the fortran compiler using "singularity exec". Note that item #2 in the list of 7 license locations is the correct one

```
"INFO:    Converting SIF file to temporary sandbox...
Abaqus License Manager checked out the following licenses:
Abaqus/Standard checked out 5 tokens from Flexnet server [server name redacted].
<145 out of 150 licenses remain available>.

Error: A license for Comp-FL is not available now (-15,570,115).

A connection to the license server could not be made.  You should
make sure that your license daemon process is running: both an
lmgrd process and an INTEL process should be running
if your license limits you to a specified number of licenses in use
at a time.  Also, check to see if the wrong port@host or the wrong
license file is being used, or if the port or hostname in the license
file has changed.

License file(s) used were (in this order):
    1.  Trusted Storage
**  2.  [port number redacted]@[server name redacted]
**  3.  /opt/intel/compilers_and_libraries_2016.1.150/linux/bin/intel64/../../Licenses
**  4.  /scratch/allocation1/workdirectory/Licenses
**  5.  /opt/intel/licenses/*.lic
**  6.  /Users/Shared/Library/Application Support/Intel/Licenses
**  7.  /opt/intel/compilers_and_libraries_2016.1.150/linux/bin/intel64/*.lic

Please visit http://software.intel.com/sites/support/ if you require technical assistance.

ifort: error #10052: could not checkout FLEXlm license
INFO:    Cleaning up image..."
```

***for third party plugin***
The following is error output when running an abaqus job requiring the third party plugin using "singularity exec".

```
INFO:    Converting SIF file to temporary sandbox...
Abaqus License Manager checked out the following licenses:
Abaqus/Standard checked out 5 tokens from Flexnet server abaqus_academic.license.crn.ubc.ca.
<145 out of 150 licenses remain available>.

*******************************************************************************
Error message from plugin:
Cannot connect to license server.
Program halted.
*******************************************************************************
INFO:    Cleaning up image...
```

**To Reproduce**
The following is the code for the PBS Batch script. This does not check out the license properly

```
#!/bin/bash
###Select resources (time, number of nodes, # of cpus per node, memory)
#PBS -l walltime=1:00:00,select=1:ncpus=1:mem=48gb

###name the job (i.e. name that shows in qstat)
#PBS -N test

###HPC account to be used
#PBS -A allocation1

###name of output and error files
#PBS -o output_multinode.file
#PBS -e error_multinode.file

###change to working directory
cd $PBS_O_WORKDIR

###load modules
module load Software_Collection/2021
module load gcc python
module load singularity # this loads singularity 3.8.5

###set path variables so that the abaqus files & intel compiler files can be found inside the singularity container
export SINGULARITYENV_PREPEND_PATH="/var/DassaultSystemes/SIMULIA/Commands/:/opt/intel/bin/"

###set environment variable for Intel Fortran Compiler license file
export SINGULARITYENV_INTEL_LICENSE_FILE="[port number redacted]@[server name redacted]"

###set environment variable for the Plugin license file
export SINGULARITYENV_PLUGIN_LICENSE_FILE="[port number redacted]@[server name redacted]"

###Set environment variable to point the plugin to the directory where the library of user subroutines is located
export SINGULARITYENV_usub_lib_dir="/var/pluginlib/linux64/abq2019"

###set nodefile location to a local directory and copy the nodefile to the local directory
export SINGULARITYENV_PBS_NODEFILE="/scratch/allocation1/workdirectory/abaqus2019singularitytest/nodefile"
cp $PBS_NODEFILE /scratch/allocation1/workdirectory/abaqus2019singularitytest/nodefile

#Execute singularity container and run abaqus job
singularity exec \
  --home /scratch/allocation1/workdirectory/abaqus2019singularitytest \
  --bind /cm/shared:/cm/shared \
  /var/project/allocation1/SingularityContainers/abaqus2019singularity_2021-12-17.sif \
  abaqus job=inputfilename cpus=1 interactive
```

The following does work as expected for the plugin (license file checks out) although it does not work for the fortran compiler. This is done from the HPC terminal and all lines are entered manually (not a script)

```
module load Software_Collection/2021
module load gcc python
module load singularity
export SINGULARITYENV_PREPEND_PATH="/var/DassaultSystemes/SIMULIA/Commands/:/opt/intel/bin/"
export SINGULARITYENV_INTEL_LICENSE_FILE="[port number redacted]@[server name redacted]"
export SINGULARITYENV_PLUGIN_LICENSE_FILE="[port number redacted]@[server name redacted]"
export SINGULARITYENV_usub_lib_dir="/var/pluginlib/linux64/abq2019"

singularity shell \
  --home /scratch/allocation1/workdirectory/abaqus2019singularitytest \
  --bind /cm/shared:/cm/shared,/scratch/allocation1/workdirectory/abaqus2019singularitytest:/mnt/dir \
  /var/project/allocation1/SingularityContainers/abaqus2019singularity_2021-12-17.sif \

*after singularity shell loads*
cd /mnt/dir
abaqus job=inputfilename cpus=1 interactive
```

**Expected behavior**
Licenses for Intel compiler and third party plugin should checkout normally regardless of whether using "singularity exec" via a pbs batch script run through the PBS job scheduler or "singularity shell" run on the login node of the HPC

**OS / Linux Distribution**
Which Linux distribution are you using?
```
for host HPC
$ cat /etc/*-release
CentOS Linux release 7.9.2009 (Core)
Cluster Manager v9.0
slave
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.9.2009 (Core)
CentOS Linux release 7.9.2009 (Core)
```

for singularity container
```
$ cat /etc/*-release
CentOS Linux release 7.9.2009 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.9.2009 (Core)
CentOS Linux release 7.9.2009 (Core)
```

**Comment by dtrudg** [comment]
Hi @scottnsbtt 

There was no SingularityCE version 3.8.5 released from this repository. The last 3.8 version released here was 3.8.4 We are now at version 3.9.4. I believe you are running the former hpcng/singularity, now apptainer/singularity 3.8.5 release from:

https://github.com/apptainer/singularity

Apologies if this is confusing... there was a fork by Sylabs in the middle of last year, and hpcng/singularity has recently renamed their project to apptainer. I'll continue here w.r.t. SingularityCE 3.9.4, but you may wish to report on the apptainer repository I'm mentioning above.

On the issue itself, this output...

```
License file(s) used were (in this order):
    1.  Trusted Storage
**  2.  [port number redacted]@[server name redacted]
```

... suggests there is no problem passing the environment into the container. You can modify your batch job to `singularity exec <other flags> /usr/bin/env` to verify everything is set as expected in the container.

Singularity uses the host network unless otherwise asked. There is no difference between `singularity exec` and `singularity shell`, or directly using the host in terms of access to the network. 

However, because `singularity exec` is running in a batch job on a cluster node, while you are using `shell` on a login node, there may be network connectivity differences between parts of the HPC system? It's quite often the case that cluster nodes are isolated from external networks more than a login node. As I understand it, the default architecture for Bright Cluster Manager uses compute nodes on a separate internal subnet, and it's possible not all external services/ports are accessible.

To troubleshoot further:

- Try the `singularity exec` instead of `singularity shell` on the login node - does it work? To rule out a bug in exec vs shell with Singularity.
- Try to launch an interactive PBS session (if permitted) and perform the lmstat shell commands in it.
- Run the lmstat commands inside a PBS job (not inside singularity). Running them on the login node does not prove that the compute nodes running the batch job have the required network connectivity.
- Look out the output of `ip a` on the login node, and a compute node / batch script to view which networks login and compute nodes are connected to. Are the compute nodes present on an  'external' network that can reach the license server, or only an internal cluster network that may require configuration to forward traffic appropriately.

**Comment by scottnsbtt** [comment]
Thanks for your response and suggestions. I did try ```singularity exec``` on the login node as suggested and it did work properly. Additionally running lmstat inside a PBS job doesn't work, so it does seem that the issue is getting the license from a compute node.

The strange thing is that the compute nodes have no trouble getting the Abaqus license from the same server, but they just can't get the other two licenses. The only difference being that the path to the Abaqus license server is specified via the "abaqus_v6.env" file located in the working directory, whereas the other license file paths are specified by setting environment variables.

I'm waiting for a colleague to return to try your other troubleshooting suggestions and will give them a shot next week.

Thanks

**Comment by scottnsbtt** [comment]
Thanks for your help @dtrudg. The issue turned out to be that the vendor daemon ports on the HPC had not been opened for traffic. Event though the license file ports were open. Both needed to be opened for the license to be checked out properly and I can confirm that this issue was not Singularity related.

Thanks
